### PR TITLE
Add stream construct

### LINKED ISSUE #20745: Streaming Constructor Implementation

### PR DESCRIPTION
**Pull request by grainier**
## Purpose
$title

Fixes #20745

## Samples
```ballerina

int i = 0;

stream<int> evenNumberStream = stream {
    i += 2;
    return { value: i };
};

var evenNumberRec = evenNumberStream.next();
int evenNumber = <int>evenNumberRec["value"];

```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/StreamConstructorNode.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.ballerinalang.model.tree.expressions;
+
+/**
+ * @since 1.2.0
+ */
+public interface StreamConstructorNode extends ExpressionNode {
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/StreamConstructorNode.java]: Add a description to the class?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -4276,6 +4278,30 @@ public void visit(BLangServiceConstructorExpr serviceConstructorExpr) {
...
+        BType targetType = ((BStreamType) streamConstructorExpr.type).constraint;
+        BType typedescType = new BTypedescType(targetType, symTable.typeDesc.tsymbol);
+        BLangTypedescExpr typedescExpr = new BLangTypedescExpr();
+        typedescExpr.resolvedType = targetType;
+        typedescExpr.type = typedescType;
+
+        BLangInvocation streamConstructInvocation = ASTBuilderUtil.createInvocationExprForMethod(
+                streamConstructorExpr.pos, symbol, new ArrayList<>(Lists.of(typedescExpr, lambdaFunction)),
+                symResolver);
+        streamConstructInvocation.type = symTable.streamType;
```
Comment (gimantha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: dont we need to set the constraint type of the stream type here?.. meaning that we have to create a new stream type, without using symTable.streamType

File: langlib/lang.stream/src/main/java/org/ballerinalang/langlib/stream/Construct.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import org.ballerinalang.jvm.values.api.BString;
+import org.ballerinalang.model.types.TypeKind;
+import org.ballerinalang.natives.annotations.Argument;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.ballerinalang.natives.annotations.ReturnType;
+
+/**
+ * Native implementation of lang.stream:construct(typeDesc, function).
+ *
+ * @since 1.2
```
Comment (gimantha) [inline langlib/lang.stream/src/main/java/org/ballerinalang/langlib/stream/Construct.java]: lets make versioning consistent with 3 digits

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 2 above)
Comment (grainier) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: At this point, the constraint would be a `pureType`.

**Pull request merged**